resources/advancements.py

```diff
@@ -0,0 +1,177 @@
+    world.advancement('glow_hunter', icon('minecraft:glow_ink_sac'), 'Mystery of the Depths', 'Kill the Octopoteuthis', 'hunter', kill_mob('tfc:octopoteuthis'), hidden=True, frame='goal')
+    world.advancement('bear_hunter', icon('tfc:large_raw_hide'), 'Bear Attack', 'Kill a Bear.', 'hunter', kill_mob('#tfc:bears'))
+    world.advancement('fishing', icon('tfc:metal/fishing_rod/copper'), 'Fisherman', 'Hook a fish with a fishing rod.', 'hunter', generic('tfc:hooked_entity', {'entity': {'type': '#tfc:small_fish'}}))
+    world.advancement('advanced_fishing', icon('tfc:metal/fishing_rod/red_steel'), 'Master Fisherman', 'Hook a dolphin or an orca with a fishing rod.', 'fishing', generic('tfc:hooked_entity', {'entity': {'type': '#tfc:needs_large_fishing_bait'}}), frame='goal')
+    world.advancement('greatest_hunter', icon('tfc:metal/javelin/red_steel'), 'Greatest Hunter', 'Hit a rabbit from 50m away with a javelin.', 'hunter', generic('tfc:stab_entity', {'entity': {'type': 'tfc:rabbit', 'distance': {'horizontal': {'min': 50}}}}))
+    world.advancement('artist', icon('minecraft:red_dye'), 'Artist', 'Procure all 16 colors of dye.', 'root', multiple(*[inventory_changed('minecraft:%s_dye' % c, name=c) for c in COLORS]), requirements=[[c] for c in COLORS], frame='goal')
+    world.advancement('familiarity', icon('tfc:food/wheat_grain'), 'A New Friend', 'Feed an animal some food to familiarize it.', 'root', generic('tfc:fed_animal', {'entity': {'type': '#tfc:animals'}}))
+    world.advancement('powderkeg', icon('tfc:powderkeg'), 'Big Boom', 'Light a powderkeg.', 'root', generic('tfc:lit', {'block': 'tfc:powderkeg'}))
+    world.advancement('full_powderkeg', icon('minecraft:gunpowder'), 'Pakratt', 'Light a fully loaded powderkeg.', 'powderkeg', generic('tfc:full_powderkeg', None))
+    world.advancement('gemologist', icon('tfc:gem/amethyst'), 'Gemologist', 'Find every gem ore in TFC.', 'nugget', multiple(*[inventory_changed('tfc:ore/%s' % g, name=g) for g in GEMS]), requirements=[[g] for g in GEMS], frame='goal')
+    world.advancement('minerologist', icon('tfc:ore/halite'), 'Minerologist', 'Find every non-metal mineral in TFC.', 'gemologist', multiple(*[inventory_changed('tfc:ore/%s' % g, name=g) for g in ALL_MINERALS]), frame='goal', requirements=[[g] for g in ALL_MINERALS])
+    world.advancement('metallurgist', icon('tfc:metal/ingot/gold'), 'Metallurgist', 'Obtain a metal-bearing specimen for every metal in TFC.', 'nugget', metal_criteria(), requirements=[[m] for m in METALS], frame='goal')
+
+def metal_criteria() -> Json:
+    criteria = dict()
+    for metal in METALS.keys():
+        if metal in ('high_carbon_red_steel', 'high_carbon_steel', 'high_carbon_black_steel', 'high_carbon_blue_steel', 'wrought_iron'):
+            criteria.update(inventory_changed('tfc:metal/ingot/%s' % metal, name=metal))
+        else:
+            criteria.update(inventory_changed({'type': 'tfc:metal_item', 'metal': 'tfc:%s' % metal}, name=metal))
+    return criteria
```

Comment:
Why does any of this exist - this seems like it would be easily solved by tags rather than this weird custom predicate type.